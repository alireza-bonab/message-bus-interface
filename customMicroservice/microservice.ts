import { NestFactory } from '@nestjs/core';
import { CustomStrategy } from '@nestjs/microservices';
import { CustomStreamingEvents } from './customStreamingEvents';

export async function createMicroservice(module: any) {
  const microservice = await NestFactory.createMicroservice<CustomStrategy>(
    module,
    {
      strategy: new CustomStreamingEvents(),
    },
  );
  return microservice;
}
