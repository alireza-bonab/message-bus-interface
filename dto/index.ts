export * from './base.dto';
export * from './saleInvoice.dto';
export * from './customer.dto';
