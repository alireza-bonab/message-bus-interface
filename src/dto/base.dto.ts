/* eslint-disable @typescript-eslint/ban-types */

export abstract class BaseType<T = unknown> {
  constructor(properties: Partial<Exclude<T, Function>> = {}) {
    Object.assign(this, properties);
  }
}

function createChannelNameFromClassName(
  className: string,
  channelNamePrefix = '',
) {
  return `${channelNamePrefix}${className
    .slice(0, 1)
    .toLowerCase()}${className.slice(1)}`;
}

export abstract class BaseDTO<T = unknown> extends BaseType<T> {
  static getChannelName(channelNamePrefix: string) {
    return createChannelNameFromClassName(
      this.prototype.constructor.name,
      channelNamePrefix,
    );
  }

  getChannelName(channelNamePrefix: string) {
    return (this as any).constructor.getChannelName(channelNamePrefix);
  }
}
