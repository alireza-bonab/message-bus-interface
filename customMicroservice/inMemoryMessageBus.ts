import { LoggerService } from '@nestjs/common';
import { validateOrReject } from 'class-validator';
import { BaseDTO } from '../dto';
import {
  getChannelNameFromEventType,
  readFromBus,
  serializeMsg,
} from './messageBus';

export type MessageHandler<T> = (message: T) => Promise<any>;

interface Subscribers {
  [key: string]: MessageHandler<any>[];
}

export class InMemoryMessageBus {
  private static _instance: InMemoryMessageBus;

  private subscribers: Subscribers = {};

  private constructor(private logger?: LoggerService) {}

  static instance(logger?: LoggerService) {
    if (!InMemoryMessageBus._instance) {
      InMemoryMessageBus._instance = new InMemoryMessageBus(logger);
    }
    return InMemoryMessageBus._instance;
  }

  // eslint-disable-next-line @typescript-eslint/no-empty-function
  close() {}

  async publish<T extends BaseDTO>(message: T, channelNamePrefix = '') {
    const channelName = getChannelNameFromEventType(message, channelNamePrefix);
    await validateOrReject(message);
    const serializedMsg = await serializeMsg(message);
    await this.handlePublish(channelName, serializedMsg);
  }

  async subscribe<T extends BaseDTO>(
    channel: string,
    messageHandler: MessageHandler<T>,
    Cls: new () => T,
  ) {
    return this.handleSubscribe(channel, async (message: any) => {
      const unSerializedMessage = await readFromBus<T>(message, Cls);
      await validateOrReject(unSerializedMessage);
      await messageHandler(unSerializedMessage);
    });
  }

  private async handlePublish(
    channel: string,
    serializedMessage: any,
  ): Promise<any> {
    if (this.subscribers[channel] && this.subscribers[channel].length) {
      try {
        await Promise.all(
          this.subscribers[channel].map((handler) => {
            return handler(serializedMessage);
          }),
        );
      } catch (e) {
        this.logger.error(e.message, e.stack);
      }
    }
  }

  private async handleSubscribe(
    channel: string,
    handler: MessageHandler<unknown>,
  ): Promise<void> {
    if (!this.subscribers[channel]) {
      this.subscribers[channel] = [];
    }
    this.subscribers[channel].push(handler);
  }
}
