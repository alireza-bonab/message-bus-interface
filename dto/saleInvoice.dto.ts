import {
  IsDate,
  IsInt,
  IsNotEmpty,
  IsOptional,
  IsString,
  Min,
} from 'class-validator';
import { BaseDTO } from './base.dto';

export class SaleInvoiceDto extends BaseDTO<SaleInvoiceDto> {
  @IsInt()
  @Min(1)
  id: number;

  @IsString()
  @IsNotEmpty()
  customerName: string;

  @IsDate()
  @IsOptional()
  createdAt: Date;
}
