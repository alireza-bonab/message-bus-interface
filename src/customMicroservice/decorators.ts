import { EventPattern } from '@nestjs/microservices';
import {
  applyDecorators,
  ArgumentsHost,
  Catch,
  ExceptionFilter,
  Logger,
  UseFilters,
} from '@nestjs/common';
import { stringify } from 'telejson';
import { BaseDTO } from '../dto';

@Catch()
class CustomEventExceptionHandler<T extends Error>
  implements ExceptionFilter<T>
{
  private logger: Logger;

  constructor() {
    this.logger = new Logger(CustomEventExceptionHandler.name);
  }

  catch(exception: T, host: ArgumentsHost): any {
    // This is an application level error, we want to log it
    const payload = stringify(host?.getArgs() || [], { maxDepth: 2 });

    this.logger.error(
      `${exception.message}\npayload: ${payload}`,
      exception.stack,
    );
  }
}

/**
 * Listen to an event published on the Custom event bus. Event listener
 * will receive an event of the same type the event is subscribed for.
 *
 * @param event
 */
export function CustomEvent<T extends BaseDTO = BaseDTO>(
  event: new () => T,
): MethodDecorator {
  return applyDecorators(
    UseFilters(CustomEventExceptionHandler),
    EventPattern<new () => T>(event),
  );
}
