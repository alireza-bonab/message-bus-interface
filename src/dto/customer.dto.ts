import {
  IsDate,
  IsInt,
  IsNotEmpty,
  IsOptional,
  IsString,
  Min,
} from 'class-validator';
import { BaseDTO } from './base.dto';

export class CustomerDto extends BaseDTO<CustomerDto> {
  @IsInt()
  @Min(1)
  id: number;

  @IsString()
  @IsNotEmpty()
  customerName: string;

  @IsDate()
  @IsOptional()
  createdAt: Date;
}
