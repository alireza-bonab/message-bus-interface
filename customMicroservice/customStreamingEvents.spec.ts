import { Test } from '@nestjs/testing';
import { Controller } from '@nestjs/common';
import { CustomerDto, SaleInvoiceDto } from '../dto';
import { CustomEvent } from './decorators';
import { createMicroservice } from './microservice';
import { InMemoryMessageBus } from './inMemoryMessageBus';
import { LoggerService } from '@nestjs/common';

describe('Server', () => {
  const messageBus = InMemoryMessageBus.instance();
  it('listens from decorators', async () => {
    let handlerInstance: ListenerControllerA;

    @Controller()
    class ListenerControllerA {
      public events: CustomerDto[] = [];

      @CustomEvent(CustomerDto)
      async handlerFunction(theEvent: CustomerDto) {
        handlerInstance = this;
        this.events.push(theEvent);
      }
    }

    const module = Test.createTestingModule({
      controllers: [ListenerControllerA],
    });

    const microservice = await createMicroservice(module);
    const listenerControllerA = microservice.get(ListenerControllerA);
    await microservice.listenAsync();

    await messageBus.publish<CustomerDto>(
      new CustomerDto({
        id: 42,
        customerName: 'James Smith',
      }),
    );

    // Check that we get the same instance of controller
    expect(listenerControllerA).toBe(handlerInstance);
    expect(listenerControllerA.events).toEqual([
      {
        id: 42,
        customerName: 'James Smith',
      },
    ]);
    await microservice.close();
  });

  it('nestjs logg the error properly', async () => {
    @Controller()
    class ListenerControllerB {
      @CustomEvent(SaleInvoiceDto)
      // eslint-disable-next-line @typescript-eslint/no-unused-vars,class-methods-use-this
      async handlerFunction(theEvent: SaleInvoiceDto) {
        return Promise.reject(new Error('boo noo'));
      }
    }

    const module = Test.createTestingModule({
      controllers: [ListenerControllerB],
    });

    const mockLogger: LoggerService = {
      error: jest.fn(),
      log: jest.fn(),
      warn: jest.fn(),
      debug: jest.fn(),
      verbose: jest.fn(),
    };

    const microservice = await createMicroservice(module);

    microservice.useLogger(mockLogger);

    await microservice.listenAsync();

    await messageBus.publish<SaleInvoiceDto>(
      new SaleInvoiceDto({
        id: 1,
        customerName: 'me',
      }),
    );

    const errMsg = `boo noo
payload: [{\"id\":1,\"customerName\":\"me\",\"_constructor-name_\":\"SaleInvoiceDto\"}]`;
    expect(mockLogger.error).toHaveBeenCalledTimes(2);
    expect(mockLogger.error).nthCalledWith(
      1,
      errMsg,
      expect.anything(),
      'CustomEventExceptionHandler',
    );

    await microservice.close();
  });

  it('publish rejects invalid message', async () => {
    await expect(
      messageBus.publish<SaleInvoiceDto>(
        new SaleInvoiceDto({
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-ignore
          id: 'this is invalid Id!',
          customerName: 'me',
        }),
      ),
    ).rejects.toMatchSnapshot();
  });
});
