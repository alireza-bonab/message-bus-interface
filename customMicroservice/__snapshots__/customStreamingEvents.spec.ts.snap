// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Server publish rejects invalid message 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isInt": "id must be an integer number",
      "min": "id must not be less than 1",
    },
    "property": "id",
    "target": SaleInvoiceDto {
      "customerName": "me",
      "id": "this is invalid Id!",
    },
    "value": "this is invalid Id!",
  },
]
`;
